src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+
+    fn get_tasks_num(&self) -> usize {
+        self.total_waiting_tasks + self.total_running_tasks
+    }
+
+    // finished_task is called when one task is finished in the
+    // thread. It will clean up the remaining information of the
+    // task in the pool.
+    fn finished_task(&mut self, group_id: T) {
+        self.total_running_tasks -= 1;
+        self.tasks.finished(group_id);
+    }
+
+    fn pop_next_task(&mut self) -> Option<Task<T>> {
+        let next_task = self.tasks.pop();
+        if next_task.is_none() {
+            return None;
+        }
+        self.total_waiting_tasks -= 1;
+        self.total_running_tasks += 1;
+        next_task
+    }
+
+    fn push_shutdown_tasks(&self, num: usize) -> Result<(), String> {
+        for _ in 0..num {
+            if let Err(e) = self.job_sender.send(false) {
+                return Err(format!("{:?}", e));
+            }
+        }
+        Ok(())
+    }
+}
+
+/// `ThreadPool` is used to execute tasks in parallel.
+/// Each task would be pushed into the pool,and when a thread
+/// is ready to process a task, it get a task from the waiting queue
+/// according the schedule queue provided in initialization.
+pub struct ThreadPool<Q, T> {
+    meta: Arc<Mutex<ThreadPoolMeta<Q, T>>>,
+    threads: Vec<JoinHandle<()>>,
+}
+
+impl<Q: ScheduleQueue<T> + Send + 'static, T: Hash + Eq + Send + Clone + 'static> ThreadPool<Q, T> {
+    pub fn new(name: String, num_threads: usize, queue: Q) -> ThreadPool<Q, T> {
+        assert!(num_threads >= 1);
+        let (tx, rx) = channel();
+        let rx = Arc::new(Mutex::new(rx));
+        let meta = Arc::new(Mutex::new(ThreadPoolMeta::new(queue, tx)));
+
+        let mut threads = Vec::with_capacity(num_threads);
+        // Threadpool threads
+        for _ in 0..num_threads {
+            threads.push(new_thread(name.clone(), rx.clone(), meta.clone()));
+        }
+
+        ThreadPool {
+            meta: meta.clone(),
+            threads: threads,
+        }
+    }
+
+    /// Executes the function `job` on a thread in the pool.
+    pub fn execute<F>(&mut self, group_id: T, job: F)
+        where F: FnOnce() + Send + 'static
+    {
+        let lock = self.meta.clone();
+        let mut meta = lock.lock().unwrap();
+        meta.push_task(group_id, job);
+    }
+
+    pub fn get_tasks_num(&self) -> usize {
+        let lock = self.meta.clone();
+        let meta = lock.lock().unwrap();
+        meta.get_tasks_num()
+    }
+
+    pub fn stop(&mut self) -> Result<(), String> {
+        {
+            let lock = self.meta.clone();
+            let tasks = lock.lock().unwrap();
+            tasks.push_shutdown_tasks(self.threads.len()).unwrap();
+        }
+        while let Some(t) = self.threads.pop() {
+            if let Err(e) = t.join() {
+                return Err(format!("{:?}", e));
+            }
+        }
+        Ok(())
+    }
+}
+
+// each thread has a worker.
+struct Worker<Q, T> {
+    job_rever: Arc<Mutex<Receiver<bool>>>,
```

Comment:
`job_rever ` is used to communicate with main thread. When a new task comes, the main thread notify `worker` to process the  new task with sending `true`. When the main thread need to  shutdown, it will send a `false` to worker to stop. @siddontang